File: client/common/protocol.go
package common

import (
	"bytes"
)

/*
	Serializes the bet object into an slice of bytes

	The variable length fields (strings) are prefixed with their length so that the receiver
	knows how much to read from the socket

	|name-length|        name       |surname-length|        surname       |document-length|        document       |birthdate-len|		birthdate   	|   bet   | lottery_id |
	|  2-bytes  | name-length bytes |    2 bytes   | surname-length bytes |    2 bytes    | document-length bytes |   2 bytes   | birthdate-length bytes| 2 bytes |   1 byte   |
*/

func serializeBet(b *Bet, lottery_id int) []byte {
	byteArrays := [][]byte{SerializeStringAndLengthToByteArray(b.PersonName),
		SerializeStringAndLengthToByteArray(b.PersonSurname),
		SerializeStringAndLengthToByteArray(b.PersonDocument),
		SerializeStringAndLengthToByteArray(b.PersonBirthDate),
		SerializeU16(b.PersonBet),
		SerializeU8(uint8(lottery_id)),
	}

	return bytes.Join(byteArrays, nil)
}

/*
	Serializes bets in chunks

	At the start of the message, we prepend 4 bytes representing the number of bets in the chunk.
	Also, a byte signaling if the server should keep listening is appended
*/
func SerializeBets(bs []Bet, lottery_id int, keep_reading bool) []byte {
	bsArrays := make([][]byte, len(bs)+2, len(bs)+2)
	bsArrays[0] = SerializeU32(uint32(len(bs)))
	for idx := range bs {
		bsArrays[1+idx] = serializeBet(&bs[idx], lottery_id)
	}
	if keep_reading {
		bsArrays[len(bsArrays)-1] = SerializeU8(uint8(1))
	} else {
		bsArrays[len(bsArrays)-1] = SerializeU8(uint8(0))
	}
	res := bytes.Join(bsArrays, nil)
	return res
}

const (
	BetStateOk = iota
	BetStateErr
)

/*
	Deserialize the bet state after the attempt of persisting it

	It reads one byte from the `readHandle` and returns the corresponding state.
*/
func DeserializeBetSavedState(readHandle func(int) ([]byte, error)) (uint8, error) {
	buffer, err := readHandle(1)
	if err != nil {

		return BetStateErr, err
	}
	if DeserializeU8(buffer) == BetStateOk {
		return BetStateOk, nil
	}
	return BetStateErr, nil
}

/*
	Deserialize the list of documents of winners

	It firts reads 4 bytes which represent how many winners there are
	Then it simply reads 2 bytes representing the length of the document
	and reads that amount of bytes from the socket to receive the document
*/
func DeserializeWinners(readHandle func(int) ([]byte, error)) ([]string, error) {
	nWinnersBuffer, err := readHandle(4)
	if err != nil {
		return nil, err
	}
	nWinners := int(DeserializeU32(nWinnersBuffer))
	winnerDocuments := make([]string, 0, nWinners)
	for i := 0; i < nWinners; i += 1 {
		lengthDocumentBuffer, err := readHandle(2)
		if err != nil {
			return nil, err
		}
		lengthDocument := DeserializeU16(lengthDocumentBuffer)
		documentBuffer, err := readHandle(int(lengthDocument))
		if err != nil {
			return nil, err
		}
		winnerDocuments = append(winnerDocuments, DeserializeString(documentBuffer))
	}
	return winnerDocuments, nil
}

File: server/common/protocol.py
from common.deserialize_utils import deserialize_bool, deserialize_u8, deserialize_u16, deserialize_u32, deserialize_string
from common.serialize_utils import serialize_u8
from common.utils import Bet
from enum import Enum
from typing import Union

def deserialize_bet(recv):
    """
    Deserializes the bet into its fields from the payload received by the client

    recv is a handle to a function wrapping `recv`
    """
    person_name_size = deserialize_u16(recv(2))
    person_name = deserialize_string(recv(person_name_size), person_name_size)

    person_surname_size = deserialize_u16(recv(2))
    person_surname = deserialize_string(recv(person_surname_size), person_surname_size)

    person_document_size = deserialize_u16(recv(2))
    person_document = deserialize_string(recv(person_document_size), person_document_size)

    person_birthdate_size = deserialize_u16(recv(2))
    person_birthdate = deserialize_string(recv(person_birthdate_size), person_birthdate_size)

    person_bet = deserialize_u16(recv(2))

    lottery_id = deserialize_u8(recv(1))
    return Bet(lottery_id, 
               person_name, 
               person_surname, 
               person_document, 
               person_birthdate, 
               person_bet)


def deserialize_bets(recv) -> Union[list[Bet], bool]:
    bets_in_chunk = deserialize_u32(recv(4))
    bets = []
    for _ in range(bets_in_chunk):
        bet = deserialize_bet(recv)
        bets.append(bet)
    keep_reading = deserialize_bool(recv(1))
    return bets, keep_reading


class SavedBetState(Enum):
    OK = 0
    ERR = 1


# The state of saving the bet is represented with a byte (0 -> Ok, 1 -> Some kind of error)
def serialize_saved_bets_status(state: SavedBetState) -> bytearray:
    """
    Serializes the state of the bet after the attempt at persisting it
    """

    return serialize_u8(state.value)
